homeassistant/components/generic/config_flow.py

```diff
@@ -257,45 +270,46 @@ async def async_test_stream(
 
     try:
         url = yarl.URL(stream_source)
-    except ValueError:
-        return {CONF_STREAM_SOURCE: "malformed_url"}
+    except ValueError as err:
+        raise InvalidStreamException("malformed_url") from err
     if not url.is_absolute():
-        return {CONF_STREAM_SOURCE: "relative_url"}
+        raise InvalidStreamException("relative_url")
     if not url.user and not url.password:
         username = info.get(CONF_USERNAME)
         password = info.get(CONF_PASSWORD)
         if username and password:
             url = url.with_user(username).with_password(password)
             stream_source = str(url)
     try:
-        stream = create_stream(
-            hass,
-            stream_source,
-            stream_options,
-            DynamicStreamSettings(),
-            "test_stream",
+        stream = PreviewStream(
+            create_stream(
+                hass,
+                stream_source,
+                stream_options,
+                DynamicStreamSettings(),
+                "test_stream",
+            )
         )
         hls_provider = stream.add_provider(HLS_PROVIDER)
-        await stream.start()
-        if not await hls_provider.part_recv(timeout=SOURCE_TIMEOUT):
-            hass.async_create_task(stream.stop())
-            return {CONF_STREAM_SOURCE: "timeout"}
-        await stream.stop()
     except StreamWorkerError as err:
```

Comment:
Does all of this error handling work?  I'm not sure that many of these exceptions are actually raised on `create_stream` given this creates a stream without starting it. Also, stream creation happens in the background inside of `_run_worker`. My impression is that it only returns an `available` bit here without providing much other feedback.

Reply:
These were added at the time of migrating to the UI.

In that case, perhaps `stream.start()` should move within the try?

The aim at the time was to get a larger number of meaningful error messages to the user, to avoid a generic 'invalid settings' message.

But I'm all for consolidating/rationalising if that's sensible here; maybe that can be a separate PR.

Reply:
I very much like the goal  of giving helpful error messages and I think it is a good idea.  However, I'm just wondering if its even possible for these errors to be thrown. I'm not sure it is possible for a `StreamWorkerError`, `PermissionError`, `OSError` etc to get raised here. My impression is none of these exceptions can be raised directly here at the moment.

Reply:
I think you are right. I just tried a few error scenarios (invalid domain, URL without a camera stream, wrong password). All of them result in specific messages in the error log/console, but give a general timeout on the UI.

So we can probably take them out. But that would be more changes than I planned here. Can it be done in a separate PR?

Reply:
Sure.